framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -285,104 +283,101 @@ protected boolean areEqualInHierarchy(
-      outsideUpperBound = ((AnnotatedWildcardType) outside).getExtendsBound();
-      outsideLowerBound = ((AnnotatedWildcardType) outside).getSuperBound();
-    } else if (TypesUtils.isCaptured(outside.getUnderlyingType())) {
-      outsideUpperBound = ((AnnotatedTypeVariable) outside).getUpperBound();
-      outsideLowerBound = ((AnnotatedTypeVariable) outside).getLowerBound();
-    } else {
-      throw new BugInCF("Expected a wildcard or captured type variable, but found " + outside);
+      areEqualVisitHistory.put(inside, outside, currentTop, result);
+      return result;
     }
-    Boolean previousResult = areEqualVisitHistory.get(inside, outside, currentTop);
-    if (previousResult != null) {
-      return previousResult;
+    if (canBeCovariant) {
+      return isSubtype(inside, outside, currentTop);
     }
-
-    areEqualVisitHistory.put(inside, outside, currentTop, true);
-    boolean result =
-        isContainedWildcard(inside, outside, outsideUpperBound, outsideLowerBound, canBeCovariant);
-    areEqualVisitHistory.put(inside, outside, currentTop, result);
-    return result;
+    return areEqualInHierarchy(inside, outside);
   }
 
-  private boolean isContainedWildcard(
+  /**
+   * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+   * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}. See
+   * {@link #isContainedBy(AnnotatedTypeMirror, AnnotatedTypeMirror, boolean)} for a full
+   * explanation.
+   *
+   * @param inside a possibly-contained type
+   * @param outsideLower the lower bound of the possibly-containing type
+   * @param outsideUpper a the upper bound of the possibly-containing type
+   * @param canBeCovariant whether or not type arguments are allowed to be covariant
+   * @return true if inside is contained by outside, or if canBeCovariant == true and {@code inside
+   *     <: outside}
+   */
+  protected boolean isContainedByBoundType(
       AnnotatedTypeMirror inside,
-      AnnotatedTypeMirror outside,
-      AnnotatedTypeMirror outsideUpperBound,
-      AnnotatedTypeMirror outsideLowerBound,
+      AnnotatedTypeMirror outsideLower,
+      AnnotatedTypeMirror outsideUpper,
       boolean canBeCovariant) {
-
-    if (inside.equals(outside)) {
-      // If they are equal, outside always contains inside.
-      return true;
-    }
-
-    if (inside.getKind() == TypeKind.WILDCARD) {
-      outsideUpperBound =
-          checker
-              .getTypeFactory()
-              .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-    }
-    while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-      if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-        return true;
+    try {
+      if (canBeCovariant) {
+        if (outsideLower.getKind() != TypeKind.NULL) {
+          return isSubtype(outsideLower, inside);
+        } else {
+          return isSubtype(inside, outsideUpper);
```

Comment:
Can there be `assert outsideUpper.getKind() != TypeKind.NULL;` here?  If so, add it -- for documentation more than for error detection.  If not, please explain.

Reply:
Upper bounds are never `TypeKind.NULL`.  I could add it, but it's nonsensical.

Reply:
Ah, that is actually the NULL type, and not an indication that the source code did not contain anything.  (Since the source code can contain a lower bound, an upper bound, or neither.)  I was incorrectly reading it as the latter.